src/compiler/ng-jest-compiler.spec.ts

```diff
@@ -1,34 +1,587 @@
+                `
+                import {Directive} from '@angular/core';
+                import {MyOtherClass} from './other-file';
+
+                @Directive()
+                export class MyDir {
+                    constructor(other: MyOtherClass) {}
+                }
+            `,
+                { emitDecoratorMetadata: false },
+            );
+
+            expect(diagnostics.length).toBe(0);
+            expect(code).toContain('const other_file_1 = require("./other-file");');
+            expect(code).toContain(dedent`
+               MyDir.ctorParameters = () => [
+                    { type: other_file_1.MyOtherClass }
+               ];
+               exports.MyDir = MyDir = tslib_1.__decorate([
+                    (0, core_1.Directive)()
+               ], MyDir);
+            `);
+        });
+
+        it.failing('should properly serialize constructor parameter with external qualified name type', () => {
+            // This test doesn't pass because `transpileModule` can't resolve `import *`
+            const { code, diagnostics } = transformCjs(`
+                import {Directive} from '@angular/core';
+                import * as externalFile from './other-file';
+
+                @Directive()
+                export class MyDir {
+                    constructor(other: externalFile.MyOtherClass) {}
+                }
+            `);
+
+            expect(diagnostics.length).toBe(0);
+            expect(code).toContain('const externalFile = require("./other-file");');
+            expect(code).toContain(dedent`
+                MyDir.ctorParameters = () => [
+                    { type: externalFile.MyOtherClass }
+                ];
+                exports.MyDir = MyDir = tslib_1.__decorate([
+                    (0, core_1.Directive)()
+                ], MyDir);
+            `);
+        });
+
+        it('should properly serialize constructor parameter with local qualified name type', () => {
+            const { code, diagnostics } = transformCjs(`
+                import {Directive} from '@angular/core';
+
+                namespace other {
+                    export class OtherClass {}
+                };
+
+                @Directive()
+                export class MyDir {
+                    constructor(other: other.OtherClass) {}
+                }
+            `);
+
+            expect(diagnostics.length).toBe(0);
+            expect(code).toContain('var other;');
+            expect(code).toContain(dedent`
+               MyDir.ctorParameters = () => [
+                    { type: other.OtherClass }
+               ];
+               exports.MyDir = MyDir = tslib_1.__decorate([
+                    (0, core_1.Directive)()
+               ], MyDir);
+            `);
+        });
+
+        it.failing('should properly downlevel constructor parameter decorators', () => {
+            // This test seems to not pass because `Document` type is unknown with `transpileModule`
```

Comment:
What if you included in the `lib.d.ts` files with `ts.transpileModule`? Would that bring `Document` into scope?

I don't see any reason TypeScript _couldn't_ know about this and still get all the isolation benefits.

Reply:
Thanks for the hint! I was missing this when reading source code of `transpileModule` and indeed it helped 😃